File: Hard/2050. Parallel Courses III/2050. Parallel Courses III.js
/**
 * @param {number} n
 * @param {number[][]} relations
 * @param {number[]} time
 * @return {number}
 */
var minimumTime = function (n, relations, time) {
    const map = new Map();
    for (const [prev, next] of relations) {
        if (map.has(next)) {
            map.get(next).push(prev);
        } else {
            map.set(next, [prev]);
        }
    }

    let max = 0;
    const memo = new Array(n + 1).fill(-1);

    for (let i = 1; i <= n; i++) {
        if (memo[i] !== -1) {
            continue;
        }
        max = Math.max(max, dfs(i));
    }

    return max;

    function dfs(node) {
        if (!map.has(node)) {
            return time[node - 1];
        }
        if (memo[node] !== -1) {
            return memo[node];
        }

        let innerMax = -1;
        const prerequisites = map.get(node);
        for (const course of prerequisites) {
            innerMax = Math.max(innerMax, time[node - 1] + dfs(course));
        }

        return (memo[node] = innerMax);
    }
};

File: Hard/2050. Parallel Courses III/2050. Parallel Courses III.py
from typing import List


class Solution:
    def minimumTime(
        self, n: int, relations: List[List[int]], time: List[int]
    ) -> int:
        map = {}
        for prev, next in relations:
            if next in map:
                map[next].append(prev)
            else:
                map[next] = [prev]

        max_time = 0
        memo = [-1] * (n + 1)

        def dfs(node):
            if node not in map:
                return time[node - 1]
            if memo[node] != -1:
                return memo[node]

            inner_max = -1
            prerequisites = map[node]
            for course in prerequisites:
                inner_max = max(inner_max, time[node - 1] + dfs(course))

            memo[node] = inner_max
            return inner_max

        for i in range(1, n + 1):
            if memo[i] == -1:
                max_time = max(max_time, dfs(i))

        return max_time
